src/v/crypto/key.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "key.h"
+
+#include "crypto/crypto.h"
+#include "crypto/exceptions.h"
+#include "crypto/ssl_utils.h"
+#include "crypto/types.h"
+
+#include <absl/container/flat_hash_set.h>
+#include <openssl/bio.h>
+#include <openssl/bn.h>
+#include <openssl/evp.h>
+#include <openssl/param_build.h>
+#include <openssl/pem.h>
+
+namespace crypto {
+
+namespace internal {
+EVP_PKEY_ptr load_public_pem_key(bytes_view key) {
+    BIO_ptr key_bio(BIO_new_mem_buf(key.data(), key.size()));
+    auto pkey_temp = EVP_PKEY_new();
+    if (
+      nullptr
+      == PEM_read_bio_PUBKEY(key_bio.get(), &pkey_temp, nullptr, nullptr)) {
+        EVP_PKEY_free(pkey_temp);
+        throw ossl_error("Failed to load PEM public key");
+    }
+
+    return EVP_PKEY_ptr(pkey_temp);
+}
+
+EVP_PKEY_ptr load_private_pem_key(bytes_view key) {
+    BIO_ptr key_bio(BIO_new_mem_buf(key.data(), key.size()));
+    auto pkey_temp = EVP_PKEY_new();
+    if (
+      nullptr
+      == PEM_read_bio_PrivateKey(key_bio.get(), &pkey_temp, nullptr, nullptr)) {
+        EVP_PKEY_free(pkey_temp);
+        throw ossl_error("Failed to load PEM private key");
+    }
+
+    return EVP_PKEY_ptr(pkey_temp);
+}
+
+static const absl::flat_hash_set<int> supported_key_types{EVP_PKEY_RSA};
+} // namespace internal
+
+key::~key() noexcept = default;
+key::impl::~impl() noexcept = default;
+
+key::impl::impl(internal::EVP_PKEY_ptr pkey, is_private_key_t is_private_key)
+  : _pkey(std::move(pkey))
+  , _is_private_key(is_private_key) {
+    auto key_type = EVP_PKEY_get_base_id(_pkey.get());
+
+    if (!internal::supported_key_types.contains(key_type)) {
+        throw exception(fmt::format("Unsupported key type {}", key_type));
+    }
+}
+
+key_type key::impl::get_key_type() const {
+    auto key_type = EVP_PKEY_get_base_id(_pkey.get());
+
+    switch (key_type) {
+    case EVP_PKEY_RSA:
+        return key_type::RSA;
+    default:
+        vassert(false, "Unsupported key type {}", key_type);
+    }
+}
+
+key_type key::get_type() const { return _impl->get_key_type(); }
+is_private_key_t key::is_private_key() const { return _impl->is_private_key(); }
+
+std::unique_ptr<key::impl>
+key::impl::load_pem_key(bytes_view key, is_private_key_t is_private_key) {
+    auto key_ptr = is_private_key == is_private_key_t::yes
+                     ? internal::load_private_pem_key(key)
+                     : internal::load_public_pem_key(key);
+
+    return std::unique_ptr<key::impl>(
+      new key::impl(std::move(key_ptr), is_private_key));
```

Comment:
The `new` actually confused me for a second, it's been so long since I've seen one.

An alternative formulation is:
```c++ 
class key::impl {
    struct _private {};

public:
    impl(
      _private p, internal::EVP_PKEY_ptr pkey, is_private_key_t is_private_key);
```

```suggestion
    return std::make_unique<key::impl>(
      _private{}, std::move(key_ptr), is_private_key);
```

Reply:
Oh I like that, thanks!